src/crush/mapper.c

```diff
@@ -319,40 +326,53 @@ static inline __s32 *get_choose_arg_ids(const struct crush_bucket_straw2 *bucket
   return arg->ids;
 }
 
+/*
+ * Compute exponential random variable using inversion method.
+ * A tiny deviation from the standard method, but for backward compatibility
+ * we keep it as it was.
+ *
+ * for reference, see the exponential distribution example at:  
+ * https://en.wikipedia.org/wiki/Inverse_transform_sampling#Examples
+ */
+static inline __s64 generate_exponential_distribution(int type, int x, int id, int r, 
+                                                      int weight)
+{
+  unsigned int u = crush_hash32_3(type, x, id, r);
+  u &= 0xffff;
+
+  /*
+   * for some reason slightly less than 0x10000 produces
+   * a slightly more accurate distribution... probably a
+   * rounding effect.
+   *
+   * the natural log lookup table maps [0,0xffff]
+   * (corresponding to real numbers [1/0x10000, 1] to
+   * [0, 0xffffffffffff] (corresponding to real numbers
+   * [-11.090355,0]).
+   */
+  __s64 ln = crush_ln(u) - 0x1000000000000ll;
+
+  /*
+   * divide by 16.16 fixed-point weight.  note
+   * that the ln value is negative, so a larger
+   * weight means a larger (less negative) value
+   * for draw.
+   */
+  return div64_s64(ln, weight);
+}
+
 static int bucket_straw2_choose(const struct crush_bucket_straw2 *bucket,
 				int x, int r, const struct crush_choose_arg *arg,
                                 int position)
 {
 	unsigned int i, high = 0;
-	unsigned int u;
-	__s64 ln, draw, high_draw = 0;
+	__s64 draw, high_draw = 0;
         __u32 *weights = get_choose_arg_weights(bucket, arg, position);
         __s32 *ids = get_choose_arg_ids(bucket, arg);
 	for (i = 0; i < bucket->h.size; i++) {
                 dprintk("weight 0x%x item %d\n", weights[i], ids[i]);
 		if (weights[i]) {
-			u = crush_hash32_3(bucket->h.hash, x, ids[i], r);
-			u &= 0xffff;
-
-			/*
-			 * for some reason slightly less than 0x10000 produces
-			 * a slightly more accurate distribution... probably a
-			 * rounding effect.
-			 *
-			 * the natural log lookup table maps [0,0xffff]
-			 * (corresponding to real numbers [1/0x10000, 1] to
-			 * [0, 0xffffffffffff] (corresponding to real numbers
-			 * [-11.090355,0]).
-			 */
-			ln = crush_ln(u) - 0x1000000000000ll;
-
-			/*
-			 * divide by 16.16 fixed-point weight.  note
-			 * that the ln value is negative, so a larger
-			 * weight means a larger (less negative) value
-			 * for draw.
-			 */
-			draw = div64_s64(ln, weights[i]);
+                        draw = generate_exponential_distribution(bucket->h.hash, x, ids[i], r, weights[i]);
```

Comment:
this change just extracts the well documented 4-lines code block into a separate function. i doubt it improves the readability.

Reply:
the original document only tells how to compute the natural log or how to divide, 
but here, the key point is exponential distribution, how to generate exponential distribution based on random(), which is evenly distribution.